File: python/src/order_book.py
from enum import Enum
from datetime import datetime
import logging_utils

class Side(str, Enum):
    BUY = "BUY"
    SELL = "SELL"


class Order:
    def __init__(self, order_id, side, shares, limit_price, entry_time):
        self.order_id = order_id
        self.side = side
        self.shares = shares
        self.limit_price = limit_price
        self.entry_time = entry_time
        self.created_time = datetime.utcnow().isoformat()
        self.filled_time = None

        # Doubly linked list pointers (within a Limit)
        self.prev_order = None
        self.next_order = None

        # Back pointer to parent Limit
        self.parent_limit = None

    def __repr__(self):
        return f"Order(id={self.order_id}, shares={self.shares})"


class Limit:
    def __init__(self, limit_price):
        self.limit_price = limit_price

        # Aggregates
        self.size = 0
        self.total_volume = 0

        # BST pointers
        self.parent = None
        self.left = None
        self.right = None

        # FIFO queue of orders
        self.head_order = None
        self.tail_order = None

    def append_order(self, order: Order):
        """Append order to FIFO queue (O(1))"""
        order.parent_limit = self

        if self.tail_order is None:
            self.head_order = self.tail_order = order
        else:
            self.tail_order.next_order = order
            order.prev_order = self.tail_order
            self.tail_order = order

        self.size += 1
        self.total_volume += order.shares

    def remove_order(self, order: Order):
        """Remove order from FIFO queue (O(1))"""
        if order.prev_order:
            order.prev_order.next_order = order.next_order
        else:
            self.head_order = order.next_order

        if order.next_order:
            order.next_order.prev_order = order.prev_order
        else:
            self.tail_order = order.prev_order

        self.size -= 1
        self.total_volume -= order.shares

        order.prev_order = None
        order.next_order = None
        order.parent_limit = None

    def __repr__(self):
        return f"Limit(price={self.limit_price}, size={self.size}, vol={self.total_volume})"


class OrderBook:
    def __init__(self):
        # Root nodes
        self.buy_tree = None
        self.sell_tree = None

        # Inside market
        self.best_bid = None
        self.best_ask = None

        # Hash maps
        self.orders_by_id = {}
        self.limits_by_price = {}

    # -------------------------
    # BST helpers
    # -------------------------

    def _bst_insert(self, root, limit, side):
        if root is None:
            return limit

        if limit.limit_price < root.limit_price:
            root.left = self._bst_insert(root.left, limit, side)
            root.left.parent = root
        else:
            root.right = self._bst_insert(root.right, limit, side)
            root.right.parent = root

        return root

    def _bst_min(self, node):
        while node.left:
            node = node.left
        return node

    def _bst_max(self, node):
        while node.right:
            node = node.right
        return node
    
    def _bst_remove(self, root, limit):
        """Remove a limit node from BST, returns new root."""
        if root is None:
            return None
        if limit.limit_price < root.limit_price:
            root.left = self._bst_remove(root.left, limit)
            if root.left:
                root.left.parent = root
        elif limit.limit_price > root.limit_price:
            root.right = self._bst_remove(root.right, limit)
            if root.right:
                root.right.parent = root
        else:
            # Found the node to remove
            if root.left is None:
                replacement = root.right
                if replacement:
                    replacement.parent = root.parent
                return replacement
            elif root.right is None:
                replacement = root.left
                if replacement:
                    replacement.parent = root.parent
                return replacement
            else:
                # Two children: replace with in-order successor, then delete successor
                successor = self._bst_min(root.right)
                root.limit_price = successor.limit_price
                root.head_order = successor.head_order
                root.tail_order = successor.tail_order
                root.size = successor.size
                root.total_volume = successor.total_volume
                # re-point orders to new parent node
                o = root.head_order
                while o:
                    o.parent_limit = root
                    o = o.next_order
                self.limits_by_price[root.limit_price] = root
                root.right = self._bst_remove(root.right, successor)
                if root.right:
                    root.right.parent = root
        return root

    # -------------------------
    # Core operations
    # -------------------------

    def add_order(self, order_id, side, shares, limit_price, entry_time):
        if order_id in self.orders_by_id:
            raise ValueError("Duplicate order ID")

        order = Order(order_id, side, shares, limit_price, entry_time)
        logging_utils.log_event(
            "ADD",
            time=order.entry_time,
            order_id=order.order_id,
            side=order.side,
            price=order.limit_price,
            qty=order.shares,
        )

        # Get or create limit
        limit = self.limits_by_price.get(limit_price)
        is_new_limit = False

        if not limit:
            limit = Limit(limit_price)
            self.limits_by_price[limit_price] = limit
            is_new_limit = True

            if side == Side.BUY:
                self.buy_tree = self._bst_insert(self.buy_tree, limit, side)
                if self.best_bid is None or limit_price > self.best_bid.limit_price:
                    self.best_bid = limit
            else:
                self.sell_tree = self._bst_insert(self.sell_tree, limit, side)
                if self.best_ask is None or limit_price < self.best_ask.limit_price:
                    self.best_ask = limit

        limit.append_order(order)
        self.orders_by_id[order_id] = order

    def cancel_order(self, order_id):
        order = self.orders_by_id.get(order_id)
        if not order:
            logging_utils.log_event("CANCEL_IGNORED", order_id=order_id)
            return

        logging_utils.log_event("CANCEL", order_id=order_id, price=order.limit_price,
            side=order.side, remaining_qty=order.shares)
        limit = order.parent_limit
        side = order.side
        limit.remove_order(order)
        del self.orders_by_id[order_id]
        self._remove_limit_if_empty(limit, side)

    def _can_match(self, limit_price, side):
        if side == Side.BUY:
            return self.best_ask and limit_price >= self.best_ask.limit_price
        else:
            return self.best_bid and limit_price <= self.best_bid.limit_price

    def submit_order(self, order_id, side, shares, limit_price, entry_time):
        """
        Tries to match new order. Keeps matching the order as long as possible
        If unsuccessful or partially filled, adds rest of the order to Order Book
        """
        while shares > 0 and self._can_match(limit_price, side):
            # resting_side needs to be inside the loop for multi-price-level orders
            # that might update best_bid/best_ask when orders at the best price run out
            resting_side = self.best_ask if side == Side.BUY else self.best_bid
            traded = min(shares, resting_side.head_order.shares)
            trade_price = resting_side.limit_price
            resting_side.head_order.shares -= traded
            shares -= traded
            if side == Side.BUY:
                buy_order_id = order_id
                sell_order_id = self.best_ask.head_order.order_id
            else:
                buy_order_id = self.best_bid.head_order.order_id
                sell_order_id = order_id

            if resting_side.head_order.shares == 0:
                del_order_id = resting_side.head_order.order_id
                resting_side.remove_order(resting_side.head_order)
                del self.orders_by_id[del_order_id]
                resting_side_side = Side.SELL if side == Side.BUY else Side.BUY
                self._remove_limit_if_empty(resting_side, resting_side_side)

            logging_utils.log_trade(
                time=datetime.utcnow().isoformat(),
                buy_order_id=buy_order_id,
                sell_order_id=sell_order_id,
                qty=traded,
                price=trade_price,
                fill='PARTIAL' if shares > 0 else 'FULL')

        if shares > 0:
            self.add_order(order_id, side, shares, limit_price, entry_time)

    def _remove_limit_if_empty(self, limit, side):
        """Remove a limit from the BST and update best_bid/best_ask if exhausted."""
        if limit.size > 0:
            return
        del self.limits_by_price[limit.limit_price]

        if side == Side.BUY:
            self.buy_tree = self._bst_remove(self.buy_tree, limit)
            # next best bid is the largest remaining value in buy_tree
            self.best_bid = self._bst_max(self.buy_tree) if self.buy_tree else None
        else:
            self.sell_tree = self._bst_remove(self.sell_tree, limit)
            # next best ask is the smallest remaining value in sell_tree
            self.best_ask = self._bst_min(self.sell_tree) if self.sell_tree else None


    # -------------------------
    # Queries
    # -------------------------

    def get_best_bid(self):
        return self.best_bid.limit_price if self.best_bid else None

    def get_best_ask(self):
        return self.best_ask.limit_price if self.best_ask else None

    def get_volume_at_price(self, price):
        limit = self.limits_by_price.get(price)
        return limit.total_volume if limit else 0

File: cpp/src/order_book.cpp
#pragma once
#include <map>
#include <list>
#include <unordered_map>
#include <optional>
#include <functional>
#include <string>
#include <chrono>

namespace ob {

enum class Side {BUY, SELL};

enum class FillType {FULL, PARTIAL};

struct Order {
    int order_id;
    Side side;
    int shares;
    int limit_price;
    double entry_time;
    std::string created_time;

    Order(int id, Side s, int sh, int price, double t)
        : order_id(id), side(s), shares(sh), limit_price(price), entry_time(t) {
            created_time = std::to_string(
            std::chrono::duration_cast<std::chrono::milliseconds>(
                std::chrono::system_clock::now().time_since_epoch()
            ).count());
        }
};

struct Trade {
    int shares;
    int buy_order_id;
    int sell_order_id;
    int price;
    FillType fill_type;
    Side incoming_side;
};

struct Limit {
    int limit_price;
    std::list<Order> orders; // FIFO queue
    int size = 0;
    long long total_volume = 0;

    explicit Limit(int p): limit_price(p) {}
};


class OrderBook {

private:
    std::map<int, Limit, std::greater<int>> bids;
    std::map<int, Limit> asks;

    struct Entry {
        Limit* limit;
        std::list<Order>::iterator it;
    };
    std::unordered_map<int, Entry> orders_by_id;

    template<typename MapType>
    void addOrderToTree(MapType &tree, int order_id, Side side, int shares, int limit_price, double entry_time) {
        auto it = tree.find(limit_price);
        if (it == tree.end()) {
            it = tree.emplace(limit_price, Limit(limit_price)).first;
        }
        Limit &lim = it->second;
        lim.orders.emplace_back(order_id, side, shares, limit_price, entry_time);
        auto list_it = std::prev(lim.orders.end());
        lim.size += 1;
        lim.total_volume += shares;
        orders_by_id.emplace(order_id, Entry{ &lim, list_it });
    }

public:
    OrderBook() = default;

    // Add order as a new Order into the book (no matching here)
    void addOrder(int order_id, Side side, int shares, int limit_price, double entry_time) {
        if (side == Side::BUY) {
            addOrderToTree(bids, order_id, side, shares, limit_price, entry_time);
        } else {
            addOrderToTree(asks, order_id, side, shares, limit_price, entry_time);
        }
    }

    // Cancel an existing order by id (no-op if not found)
    void cancelOrder(int order_id) {
        auto o_it = orders_by_id.find(order_id);
        if (o_it == orders_by_id.end()) return;
        Limit *lim = o_it->second.limit;
        auto list_it = o_it->second.it;
        lim->total_volume -= list_it->shares;
        lim->size -= 1;
        lim->orders.erase(list_it);
        orders_by_id.erase(o_it);

        // remove empty limit from its tree
        removeLimitIfEmpty(lim->limit_price);
    }


    // Submit order: try to match immediately; if any remainder, add to book
    // Returns vector of Trades executed (may be empty)
    std::vector<Trade> submitOrder(int order_id, Side side, int shares, int limit_price, double entry_time) {
        std::vector<Trade> trades;
        int remaining = shares;
        // keep executing against the inside until order fully filled or no crossing
        while (remaining > 0) {
            Limit *restingLimit = (side == Side::BUY) ? bestAskLimit() : bestBidLimit();
            // break if no match can be made
            if (!restingLimit) break;
            if (side == Side::BUY && restingLimit->limit_price > limit_price) break;
            if (side == Side::SELL && restingLimit->limit_price < limit_price) break;

            Order &oppOrder = restingLimit->orders.front();
            int take = std::min(remaining, oppOrder.shares);
            Trade tr;
            if (side == Side::BUY) {
                tr.buy_order_id = order_id;
                tr.sell_order_id = oppOrder.order_id;
            } else {
                tr.buy_order_id = oppOrder.order_id;
                tr.sell_order_id = order_id;
            }
            tr.price = restingLimit->limit_price;
            tr.shares = take;
            tr.incoming_side = side;
            // apply trade to book (reuse applyTrade helper)
            applyTrade(restingLimit, take, oppOrder.order_id);
            remaining -= take;
            tr.fill_type = (remaining == 0) ? FillType::FULL : FillType::PARTIAL;
            trades.push_back(tr);
        }

        if (remaining > 0) {
            // add the remaining as a resting order
            addOrder(order_id, side, remaining, limit_price, entry_time);
        }
        return trades;
    }

    // Execute one best-against-best match, return Trade if a match happened
    std::optional<Trade> executeBest() {
        auto bidIt = bids.begin();
        auto askIt = asks.begin();
        if (bidIt == bids.end() || askIt == asks.end()) return std::nullopt;
        if (bidIt->first < askIt->first) return std::nullopt; // no crossing

        Limit &buyL = bidIt->second;
        Limit &sellL = askIt->second;
        if (buyL.orders.empty() || sellL.orders.empty()) return std::nullopt;

        Order &buyOrder = buyL.orders.front();
        Order &sellOrder = sellL.orders.front();

        int traded = std::min(buyOrder.shares, sellOrder.shares);
        Trade tr;
        tr.shares = traded;
        tr.buy_order_id = buyOrder.order_id;
        tr.sell_order_id = sellOrder.order_id;
        // price convention: use sell (aggressed) price
        tr.price = sellL.limit_price;

        // apply fills
        applyTrade(&buyL, traded, buyOrder.order_id);
        applyTrade(&sellL, traded, sellOrder.order_id);

        return tr;
    }

    // Queries
    std::optional<int> bestBid() const {
        if (bids.empty()) return std::nullopt;
        return bids.begin()->first;
    }
    std::optional<int> bestAsk() const {
        if (asks.empty()) return std::nullopt;
        return asks.begin()->first;
    }
    

private:
    void removeLimitIfEmpty(int price) {
        auto b = bids.find(price);
        if (b != bids.end() && b->second.orders.empty()) {
            bids.erase(b);
            return;
        }
        auto a = asks.find(price);
        if (a != asks.end() && a->second.orders.empty()) {
            asks.erase(a);
        }
    }

    Limit* bestAskLimit() {
        if (asks.empty()) return nullptr;
        return &asks.begin()->second;
    }
    Limit* bestBidLimit() {
        if (bids.empty()) return nullptr;
        return &bids.begin()->second;
    }

    // apply trade quantity to a specific limit's front order (identified by id)
    // This reduces shares and removes order+limit as necessary.
    void applyTrade(Limit* lim, int qty, int order_id) {
        if (!lim || lim->orders.empty()) return;
        Order &o = lim->orders.front();
        if (o.order_id != order_id) {
            // find via orders_by_id (shouldn't happen if caller identified correctly)
            auto it = orders_by_id.find(order_id);
            if (it == orders_by_id.end()) return;
            lim = it->second.limit;
            auto list_it = it->second.it;
            Order &oo = *list_it;
            int use = std::min(qty, oo.shares);
            oo.shares -= use;
            lim->total_volume -= use;
            if (oo.shares == 0) {
                lim->orders.erase(list_it);
                orders_by_id.erase(it);
            }
            removeLimitIfEmpty(lim->limit_price);
            return;
        }

        int use = std::min(qty, o.shares);
        o.shares -= use;
        lim->total_volume -= use;
        if (o.shares == 0) {
            // remove front
            orders_by_id.erase(o.order_id);
            lim->orders.pop_front();
            lim->size -= 1;
        }
        removeLimitIfEmpty(lim->limit_price);
    }
};

}